{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Custom Report Generator\n",
    "### Takes in tNGS outfile and sequencing load file to create 'custom report' file to be able to import into Starlims"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import os\n",
    "import time"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>Well</td>\n",
       "      <td>Sample Name</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>5</td>\n",
       "      <td>A01</td>\n",
       "      <td>cnvcontrol01_F_SF2001234</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>6</td>\n",
       "      <td>B01</td>\n",
       "      <td>cnvcontrol01_M_SF2001234</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>7</td>\n",
       "      <td>C01</td>\n",
       "      <td>EX1100630_SF2001234</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>8</td>\n",
       "      <td>D01</td>\n",
       "      <td>EX1702245_SF2001234</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>100</td>\n",
       "      <td>F04</td>\n",
       "      <td>EX1816211_SF2001234</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>101</td>\n",
       "      <td>G04</td>\n",
       "      <td>EX1816212_SF2001234</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>102</td>\n",
       "      <td>H04</td>\n",
       "      <td>EX1816215_SF2001234</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>103</td>\n",
       "      <td>A05</td>\n",
       "      <td>EX1816215_SF2001234</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>104</td>\n",
       "      <td>B05</td>\n",
       "      <td>snvcontrol01_SF2001234</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>101 rows Ã— 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        0                         1\n",
       "4    Well               Sample Name\n",
       "5     A01  cnvcontrol01_F_SF2001234\n",
       "6     B01  cnvcontrol01_M_SF2001234\n",
       "7     C01       EX1100630_SF2001234\n",
       "8     D01       EX1702245_SF2001234\n",
       "..    ...                       ...\n",
       "100   F04       EX1816211_SF2001234\n",
       "101   G04       EX1816212_SF2001234\n",
       "102   H04       EX1816215_SF2001234\n",
       "103   A05       EX1816215_SF2001234\n",
       "104   B05    snvcontrol01_SF2001234\n",
       "\n",
       "[101 rows x 2 columns]"
      ]
     },
     "execution_count": 49,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tngs = pd.read_csv('../output/tNGS_export.csv')\n",
    "# seq = pd.read_csv('../data/seq1.csv',delimiter=\"\\t\")\n",
    "seq = pd.read_csv('../data/seq2.csv',delimiter=\"\\t\", header=None)\n",
    "seq = seq.iloc[4:,:2].copy()\n",
    "tngs.columns = [c.replace(' ', '_') for c in tngs.columns]\n",
    "#seq.columns = [c.replace(' ', '_') for c in seq.columns]\n",
    "seq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Starlims sequencing'NGS' workbatch number\n",
    "workbatchNo = seq['Container_Name'][0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get starlims 'amplicon'\n",
    "star_amplicon = \"_NGS_\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "# need to get a dictionary of all sample ID from sequencing load file as that represents what can get \n",
    "# imported back into Starlims. Mostly important for the ddPCR mutation details import\n",
    "samples = list(seq['Plate_ID'].unique())\n",
    "del samples[:4]\n",
    "samples = [sample[:9] for sample in samples]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "# new column with sample id and mutsurveyor variant\n",
    "tngs['id_variant'] = tngs['Folder_number'] + \"_\" + tngs['MutSurveyor']\n",
    "tngs_var = list(tngs['id_variant'].unique())\n",
    "tngs_var = [var for var in tngs_var if str(var) != 'nan']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "# new column with sample id and mutation details variant\n",
    "tngs['id_mut'] = tngs['Folder_number'] + \"_\" + tngs['MutDetails']\n",
    "tngs_mut = list(tngs['id_mut'].unique())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "# new columns with sample id and genomic nomenclature\n",
    "tngs['id_genomic'] = tngs['Folder_number'] + \"_\" + tngs['Genomic_nomenclature']\n",
    "tngs_genomic = list(tngs['id_genomic'].unique())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "# create a dictionary with sample id's and associated variants\n",
    "sample_vars = {k:[] for k in samples}\n",
    "for sample in samples:\n",
    "    for variant in tngs_var:\n",
    "        if sample == variant[:9]:\n",
    "            try:\n",
    "                sample_vars[sample].append(variant[10:])\n",
    "            except:\n",
    "                sample_vars[sample] = [variant[10:]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "# create custom report template\n",
    "def create_custom_report():\n",
    "    f = open(f'{os.getcwd()}/../output/custom_report.txt', 'w+')\n",
    "    header = \"Warning!\\nSample Name\\tReference Name\\tLane Quality\\tROI Coverage\\t#nts below threshold\\tQuality ROI\\tVariant1\\tVariant3\\tVariant3\\tVariant4\\n\"\n",
    "    body = \"\"\n",
    "    for sample in sample_vars:\n",
    "        no_vars = len(sample_vars[sample])\n",
    "        sample_id = sample + star_amplicon + workbatchNo\n",
    "        if no_vars == 1:\n",
    "            body += f\"{sample_id}\\t\\t\\t\\t\\t\\t{sample_vars[sample][0]}\\n\"\n",
    "        elif no_vars == 2:\n",
    "            body += f\"{sample_id}\\t\\t\\t\\t\\t\\t{sample_vars[sample][0]}\\t{sample_vars[sample][1]}\\n\"        \n",
    "        elif no_vars == 3:\n",
    "            body += f\"{sample_id}\\t\\t\\t\\t\\t\\t{sample_vars[sample][0]}\\t{sample_vars[sample][1]}\\t{sample_vars[sample][2]}\\n\" \n",
    "        elif no_vars == 4:\n",
    "            body += f\"{sample_id}\\t\\t\\t\\t\\t\\t{sample_vars[sample][0]}\\t{sample_vars[sample][1]}\\t{sample_vars[sample][2]}\\t{sample_vars[sample][3]}\\n\"\n",
    "        else:\n",
    "            body += f\"{sample_id}\\t\\t\\t\\t\\t\\t\\t\\t\\t\\t\\n\"\n",
    "    f.write(header)\n",
    "    f.write(body)\n",
    "    f.close()\n",
    "\n",
    "create_custom_report()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "# create mutation details 'ddPCR' import file\n",
    "\n",
    "# cols - well, sampleID, variant\n",
    "\n",
    "def create_mut_details():\n",
    "    md_import = seq.iloc[3:,:].copy()\n",
    "\n",
    "create_mut_details()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Well</th>\n",
       "      <th>Sample_Name</th>\n",
       "      <th>well_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>A01</td>\n",
       "      <td>cnvcontrol01_F_SF2001234</td>\n",
       "      <td>A01_cnvcontro</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>B01</td>\n",
       "      <td>cnvcontrol01_M_SF2001234</td>\n",
       "      <td>B01_cnvcontro</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>C01</td>\n",
       "      <td>EX1100630_SF2001234</td>\n",
       "      <td>C01_EX1100630</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>D01</td>\n",
       "      <td>EX1702245_SF2001234</td>\n",
       "      <td>D01_EX1702245</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>E01</td>\n",
       "      <td>EX1714297_SF2001234</td>\n",
       "      <td>E01_EX1714297</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>95</td>\n",
       "      <td>F04</td>\n",
       "      <td>EX1816211_SF2001234</td>\n",
       "      <td>F04_EX1816211</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>96</td>\n",
       "      <td>G04</td>\n",
       "      <td>EX1816212_SF2001234</td>\n",
       "      <td>G04_EX1816212</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>97</td>\n",
       "      <td>H04</td>\n",
       "      <td>EX1816215_SF2001234</td>\n",
       "      <td>H04_EX1816215</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>98</td>\n",
       "      <td>A05</td>\n",
       "      <td>EX1816215_SF2001234</td>\n",
       "      <td>A05_EX1816215</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>99</td>\n",
       "      <td>B05</td>\n",
       "      <td>snvcontrol01_SF2001234</td>\n",
       "      <td>B05_snvcontro</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>100 rows Ã— 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   Well               Sample_Name        well_id\n",
       "0   A01  cnvcontrol01_F_SF2001234  A01_cnvcontro\n",
       "1   B01  cnvcontrol01_M_SF2001234  B01_cnvcontro\n",
       "2   C01       EX1100630_SF2001234  C01_EX1100630\n",
       "3   D01       EX1702245_SF2001234  D01_EX1702245\n",
       "4   E01       EX1714297_SF2001234  E01_EX1714297\n",
       "..  ...                       ...            ...\n",
       "95  F04       EX1816211_SF2001234  F04_EX1816211\n",
       "96  G04       EX1816212_SF2001234  G04_EX1816212\n",
       "97  H04       EX1816215_SF2001234  H04_EX1816215\n",
       "98  A05       EX1816215_SF2001234  A05_EX1816215\n",
       "99  B05    snvcontrol01_SF2001234  B05_snvcontro\n",
       "\n",
       "[100 rows x 3 columns]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "md = seq.iloc[3:,:].copy()\n",
    "md = md.iloc[:,:2].copy()\n",
    "md.to_csv(\"md.csv\", header=None, index=False)\n",
    "md = pd.read_csv(\"md.csv\")\n",
    "md.columns = [c.replace(' ', '_') for c in md.columns]\n",
    "md[\"well_id\"] = md[\"Well\"] + \"_\" + md[\"Sample_Name\"].str[:9]\n",
    "md"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Mutation details import\n",
    "- Well | SampleID | MutDetail | Genomic\n",
    "- Well + SampleID taken from seq.csv\n",
    "- MutDetail and Genomic taken from tNGS_export.csv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['Sample Name',\n",
       " 'cnvcontrol01_F_SF2001234',\n",
       " 'cnvcontrol01_M_SF2001234',\n",
       " 'EX1100630_SF2001234',\n",
       " 'EX1702245_SF2001234',\n",
       " 'EX1714297_SF2001234',\n",
       " 'EX1806104_SF2001234',\n",
       " 'EX1806937_SF2001234',\n",
       " 'EX1809676_SF2001234',\n",
       " 'EX1812034_SF2001234',\n",
       " 'EX1812591_SF2001234',\n",
       " 'EX1812591_SF2001234',\n",
       " 'EX1813292_SF2001234',\n",
       " 'EX1813352_SF2001234',\n",
       " 'EX1813353_SF2001234',\n",
       " 'EX1813386_SF2001234',\n",
       " 'EX1813403_SF2001234',\n",
       " 'EX1813529_SF2001234',\n",
       " 'EX1813576_SF2001234',\n",
       " 'EX1813608_SF2001234',\n",
       " 'EX1813656_SF2001234',\n",
       " 'EX1813905_SF2001234',\n",
       " 'EX1813935_SF2001234',\n",
       " 'EX1813937_SF2001234',\n",
       " 'EX1813996_SF2001234',\n",
       " 'EX1814024_SF2001234',\n",
       " 'EX1814026_SF2001234',\n",
       " 'EX1814264_SF2001234',\n",
       " 'EX1814312_SF2001234',\n",
       " 'EX1814341_SF2001234',\n",
       " 'EX1814400_SF2001234',\n",
       " 'EX1814483_SF2001234',\n",
       " 'EX1814502_SF2001234',\n",
       " 'EX1814537_SF2001234',\n",
       " 'EX1814546_SF2001234',\n",
       " 'EX1814680_SF2001234',\n",
       " 'EX1814711_SF2001234',\n",
       " 'EX1814713_SF2001234',\n",
       " 'EX1814825_SF2001234',\n",
       " 'EX1814881_SF2001234',\n",
       " 'EX1814956_SF2001234',\n",
       " 'EX1814967_SF2001234',\n",
       " 'EX1814972_SF2001234',\n",
       " 'EX1815020_SF2001234',\n",
       " 'EX1815070_SF2001234',\n",
       " 'EX1815231_SF2001234',\n",
       " 'EX1815289_SF2001234',\n",
       " 'EX1815303_SF2001234',\n",
       " 'EX1815458_SF2001234',\n",
       " 'EX1815468_SF2001234',\n",
       " 'EX1815481_SF2001234',\n",
       " 'EX1815526_SF2001234',\n",
       " 'EX1815533_SF2001234',\n",
       " 'EX1815574_SF2001234',\n",
       " 'EX1815576_SF2001234',\n",
       " 'EX1815589_SF2001234',\n",
       " 'EX1815592_SF2001234',\n",
       " 'EX1815601_SF2001234',\n",
       " 'EX1815610_SF2001234',\n",
       " 'EX1815652_SF2001234',\n",
       " 'EX1815656_SF2001234',\n",
       " 'EX1815669_SF2001234',\n",
       " 'EX1815698_SF2001234',\n",
       " 'EX1815704_SF2001234',\n",
       " 'EX1815720_SF2001234',\n",
       " 'EX1815815_SF2001234',\n",
       " 'EX1815821_SF2001234',\n",
       " 'EX1815868_SF2001234',\n",
       " 'EX1815869_SF2001234',\n",
       " 'EX1815886_SF2001234',\n",
       " 'EX1815890_SF2001234',\n",
       " 'EX1815915_SF2001234',\n",
       " 'EX1815935_SF2001234',\n",
       " 'EX1815936_SF2001234',\n",
       " 'EX1815937_SF2001234',\n",
       " 'EX1815942_SF2001234',\n",
       " 'EX1815946_SF2001234',\n",
       " 'EX1815950_SF2001234',\n",
       " 'EX1815971_SF2001234',\n",
       " 'EX1815980_SF2001234',\n",
       " 'EX1816017_SF2001234',\n",
       " 'EX1816039_SF2001234',\n",
       " 'EX1816113_SF2001234',\n",
       " 'EX1816116_SF2001234',\n",
       " 'EX1816118_SF2001234',\n",
       " 'EX1816119_SF2001234',\n",
       " 'EX1816141_SF2001234',\n",
       " 'EX1816150_SF2001234',\n",
       " 'EX1816156_SF2001234',\n",
       " 'EX1816158_SF2001234',\n",
       " 'EX1816159_SF2001234',\n",
       " 'EX1816170_SF2001234',\n",
       " 'EX1816179_SF2001234',\n",
       " 'EX1816181_SF2001234',\n",
       " 'EX1816211_SF2001234',\n",
       " 'EX1816211_SF2001234',\n",
       " 'EX1816211_SF2001234',\n",
       " 'EX1816212_SF2001234',\n",
       " 'EX1816215_SF2001234',\n",
       " 'EX1816215_SF2001234',\n",
       " 'snvcontrol01_SF2001234']"
      ]
     },
     "execution_count": 70,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "list(seq[1])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create Well-Sample dictionary {well:Sample} from seq.csv\n",
    "well_sample = {}\n",
    "well = [x for x in seq[0].unique() if len(x) == 3]\n",
    "seqID = [str(x)[:9] for x in seq[1].unique()]\n",
    "for w, s in zip(well, seqID):\n",
    "    well_sample[w] = s"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'EX1100630': [['nan exnan p. g.'], []],\n",
       " 'EX1702245': [['nan exnan p. g.'], []],\n",
       " 'EX1714297': [['nan exnan p. g.'], []],\n",
       " 'EX1806104': [['nan exnan p. g.'], []],\n",
       " 'EX1806937': [['nan exnan p. g.'], []],\n",
       " 'EX1809676': [['nan exnan p. g.'], []],\n",
       " 'EX1812034': [['nan exnan p. g.'], []],\n",
       " 'EX1812591': [['GLIS3 ex4 p.Gln528Ter/N c.1582C>T/N',\n",
       "   'GLIS3 ex4 p.His520Asn/N c.1558C>A/N'],\n",
       "  ['Chr9(GRCh37):g.4117896G>A', 'Chr9(GRCh37):g.4117920G>T']],\n",
       " 'EX1813292': [['nan exnan p. g.'], []],\n",
       " 'EX1813352': [['nan exnan p. g.'], []],\n",
       " 'EX1813353': [['nan exnan p. g.'], []],\n",
       " 'EX1813386': [['GATA6 ex2 p.Pro63Ser/N c.187C>T/N'],\n",
       "  ['Chr18(GRCh37):g.19751292C>T']],\n",
       " 'EX1813403': [['nan exnan p. g.'], []],\n",
       " 'EX1813529': [['nan exnan p. g.'], []],\n",
       " 'EX1813576': [['nan exnan p. g.'], []],\n",
       " 'EX1813608': [['CASR ex7 p.Cys787Arg/N c.2359T>C/N'],\n",
       "  ['Chr3(GRCh37):g.122003160T>C']],\n",
       " 'EX1813656': [['nan exnan p. g.'], []],\n",
       " 'EX1813905': [['HNF4A ex8 p.Arg311Cys/N c.931C>T/N'],\n",
       "  ['Chr20(GRCh37):g.43052762C>T']],\n",
       " 'EX1813935': [['HNF4A ex8 p.Leu319dup/N c.956_958dup/N'],\n",
       "  ['Chr20(GRCh37):g.43052787_43052789dup']],\n",
       " 'EX1813937': [['nan exnan p. g.'], []],\n",
       " 'EX1813996': [['nan exnan p. g.'], []],\n",
       " 'EX1814024': [['nan exnan p. g.'], []],\n",
       " 'EX1814026': [['nan exnan p. g.'], []],\n",
       " 'EX1814264': [['nan exnan p. g.'], []],\n",
       " 'EX1814312': [['nan exnan p. g.'], []],\n",
       " 'EX1814341': [['INSR ex12 p.Thr771Met/N c.2312C>T/N'],\n",
       "  ['Chr19(GRCh37):g.7143057G>A']],\n",
       " 'EX1814400': [['nan exnan p. g.'], []],\n",
       " 'EX1814483': [['nan exnan p. g.'], []],\n",
       " 'EX1814502': [['nan exnan p. g.'], []],\n",
       " 'EX1814537': [['nan exnan p. g.'], []],\n",
       " 'EX1814546': [['nan exnan p. g.'], []],\n",
       " 'EX1814680': [['GCK ex5 p.Arg191Trp/N c.571C>T/N'],\n",
       "  ['Chr7(GRCh37):g.44189576G>A']],\n",
       " 'EX1814711': [['nan exnan p. g.'], []],\n",
       " 'EX1814713': [['nan exnan p. g.'], []],\n",
       " 'EX1814825': [['nan exnan p. g.'], []],\n",
       " 'EX1814881': [['nan exnan p. g.'], []],\n",
       " 'EX1814956': [['nan exnan p. g.'], []],\n",
       " 'EX1814967': [['nan exnan p. g.'], []],\n",
       " 'EX1814972': [['nan exnan p. g.'], []],\n",
       " 'EX1815020': [['nan exnan p. g.'], []],\n",
       " 'EX1815070': [['nan exnan p. g.'], []],\n",
       " 'EX1815231': [['HNF1A ex4 p.Glu275del/N c.824_826del/N'],\n",
       "  ['Chr12(GRCh37):g.121432077_121432079del']],\n",
       " 'EX1815289': [['ABCC8 ex37 p.Gly1479Arg/N c.4435G>A/N'],\n",
       "  ['Chr11(GRCh37):g.17415926C>T']],\n",
       " 'EX1815303': [['nan exnan p. g.'], []],\n",
       " 'EX1815458': [['nan exnan p. g.'], []],\n",
       " 'EX1815468': [['CASR ex2 p.Arg25Ter/N c._000388.3C>T/N'],\n",
       "  ['Chr3(GRCh37):g.121973109C>T']],\n",
       " 'EX1815481': [['nan exnan p. g.'], []],\n",
       " 'EX1815526': [['nan exnan p. g.'], []],\n",
       " 'EX1815533': [['nan exnan p. g.'], []],\n",
       " 'EX1815574': [['nan exnan p. g.'], []],\n",
       " 'EX1815576': [['CASR ex7 p.Gln983ArgfsTer/N c.2946del/N'],\n",
       "  ['Chr3(GRCh37):g.122003747del']],\n",
       " 'EX1815589': [['nan exnan p. g.'], []],\n",
       " 'EX1815592': [['nan exnan p. g.'], []],\n",
       " 'EX1815601': [['nan exnan p. g.'], []],\n",
       " 'EX1815610': [['nan exnan p. g.'], []],\n",
       " 'EX1815652': [['nan exnan p. g.'], []],\n",
       " 'EX1815656': [['SDHC ex5 p./N g.161326349-161326708del/N'],\n",
       "  ['chr1:161326349-161326708']],\n",
       " 'EX1815669': [['nan exnan p. g.'], []],\n",
       " 'EX1815698': [['HNF1A ex2 p.Trp165Ter/N c.494G>A/N'],\n",
       "  ['Chr12(GRCh37):g.121426803G>A']],\n",
       " 'EX1815704': [['nan exnan p. g.'], []],\n",
       " 'EX1815720': [['nan exnan p. g.'], []],\n",
       " 'EX1815815': [['nan exnan p. g.'], []],\n",
       " 'EX1815821': [['MT-TL1 ex1 p./N c.L1:14A>G/N'], ['NC_012920.1:g.3243A>G']],\n",
       " 'EX1815868': [['nan exnan p. g.'], []],\n",
       " 'EX1815869': [['PHEX ex2 p. g.22056421-22056780del'],\n",
       "  ['chrX:22056421-22056780']],\n",
       " 'EX1815886': [['TG ex25 p.Val1664Met/N c.4990G>A/N'],\n",
       "  ['Chr8(GRCh37):g.133948058G>A']],\n",
       " 'EX1815890': [['HNF4A ex2 p.Arg67Gln/N c.200G>A/N'],\n",
       "  ['Chr20(GRCh37):g.43034848G>A']],\n",
       " 'EX1815915': [['ABCC8 ex13 p.Glu612Lys/N c.1834G>A/N'],\n",
       "  ['Chr11(GRCh37):g.17450201C>T']],\n",
       " 'EX1815935': [['nan exnan p. g.'], []],\n",
       " 'EX1815936': [['nan exnan p. g.'], []],\n",
       " 'EX1815937': [['nan exnan p. g.'], []],\n",
       " 'EX1815942': [['nan exnan p. g.'], []],\n",
       " 'EX1815946': [['nan exnan p. g.'], []],\n",
       " 'EX1815950': [['nan exnan p. g.'], []],\n",
       " 'EX1815971': [['nan exnan p. g.'], []],\n",
       " 'EX1815980': [['HNF1A ex4 p.Pro291Gln/N c.872C>A/N'],\n",
       "  ['Chr12(GRCh37):g.121432125C>A']],\n",
       " 'EX1816017': [['nan exnan p. g.'], []],\n",
       " 'EX1816039': [['nan exnan p. g.'], []],\n",
       " 'EX1816113': [['nan exnan p. g.'], []],\n",
       " 'EX1816116': [['nan exnan p. g.'], []],\n",
       " 'EX1816118': [['SDHA ex2 p.Asp49Gly/N c.146A>G/N'],\n",
       "  ['Chr5(GRCh37):g.223679A>G']],\n",
       " 'EX1816119': [['nan exnan p. g.'], []],\n",
       " 'EX1816141': [['CASR ex7 p.Pro748Arg/N c.2243C>G/N'],\n",
       "  ['Chr3(GRCh37):g.122003044C>G']],\n",
       " 'EX1816150': [['HNF1A ex7 p.Pro447Leu/N c.1340C>T/N'],\n",
       "  ['Chr12(GRCh37):g.121435307C>T']],\n",
       " 'EX1816156': [['nan exnan p. g.'], []],\n",
       " 'EX1816158': [['nan exnan p. g.'], []],\n",
       " 'EX1816159': [['nan exnan p. g.'], []],\n",
       " 'EX1816170': [['PHEX ex4 p. g.22094349-22094708del'],\n",
       "  ['chrX:22094349-22094708']],\n",
       " 'EX1816179': [['nan exnan p. g.'], []],\n",
       " 'EX1816181': [['nan exnan p. g.'], []],\n",
       " 'EX1816211': [['TBX6 ex6 p.Arg272Gln/N c.815G>A/N',\n",
       "   'TBX6 ex4 p.Gly162Ser/N c.484G>A/N',\n",
       "   'TBX6 ex4 p.Arg119His/N c.356G>A/N'],\n",
       "  ['Chr16(GRCh37):g.30099890C>T',\n",
       "   'Chr16(GRCh37):g.30100401C>T',\n",
       "   'Chr16(GRCh37):g.30100529C>T']],\n",
       " 'EX1816212': [['nan exnan p. g.'], []],\n",
       " 'EX1816215': [['SLC34A3 ex8 p.Thr267Pro/N c.799A>C/N',\n",
       "   'PHEX ex12 p.Trp456Cys c.1368G>T'],\n",
       "  ['Chr9(GRCh37):g.140128127A>C', 'ChrX(GRCh37):g.22151705G>T']],\n",
       " 'Sample Na': [[], []],\n",
       " 'cnvcontro': [[], []],\n",
       " 'snvcontro': [[], []]}"
      ]
     },
     "execution_count": 48,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# new seq dictionary, nested lists with mutdetails and genomic nomenclature\n",
    "seq_mut = {k:[[],[]] for k in seqID}\n",
    "for sample in seqID:\n",
    "    for variant in tngs_mut:\n",
    "        if sample == variant[:9]:\n",
    "            try:\n",
    "                seq_mut[sample][0].append(variant[10:])\n",
    "            except:\n",
    "                seq_mut[sample][0] = [variant[10:]]\n",
    "    for genomic in tngs_genomic:\n",
    "        if sample == str(genomic)[:9]:\n",
    "            try:\n",
    "                seq_mut[sample][1].append(str(genomic)[10:])\n",
    "            except:\n",
    "                seq_mut[sample][1] = [str(genomic)[10:]]\n",
    "seq_mut"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "97"
      ]
     },
     "execution_count": 68,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(seqID)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
