{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# tNGS Starlims Import\n",
    "\n",
    "## Mutation Surveyor import formats\n",
    "- SNV = c.[x]+[y/=],p.A123A/B\n",
    "- Structural variants = c.123_456_hetdup/hetdel/hetinsACTG or c.123_456_del/dup/ins\n",
    "\n",
    "## Extra Columns\n",
    "- Accession number\n",
    "- Single gene\n",
    "- NucleotideNom\n",
    "- AminoNom\n",
    "- GenomicNom\n",
    "\n",
    "## Pattern logic\n",
    "- If SNV or structural variant\n",
    "- Genotype\n",
    "\n",
    "## Column for mutation details import\n",
    "- GENE p./N c./N"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import os\n",
    "import re\n",
    "# df_all = pd.read_csv('../data/tNGS_import.csv')\n",
    "df_all = pd.read_csv('../data/2018-11-09_2861_variants.12-11-2018_STARLiMS_import.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get ride of unwanted columns and create a new df\n",
    "df = df_all.iloc[:,:23].copy()\n",
    "df.rename(columns={'WT nucleotides':'Ref', 'Variant nucleotides':'Alt', 'Variant type':'varType', 'Inserted nucleotides':'insBases', 'Report variant?':'reportVar'}, inplace=True)\n",
    "sample_ids = list(x for x in df['Folder number'] if x.startswith(\"EX\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# SYstem functions\n",
    "\n",
    "def export_csv(outdir, outname):\n",
    "    if not os.path.exists(outdir):\n",
    "        os.mkdir(outdir)\n",
    "    \n",
    "    fullname = os.path.join(outdir, outname)\n",
    "    df.to_csv(fullname)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# AminoAcid - Add more code to give the final nomenclature (e.g. replace =, * with ref amino acid and Ter)\n",
    "def amino_acid(string):\n",
    "    amino_acid = str(re.findall(\"[a-zA-Z]{3}[0-9]+[\\_]*[a-zA-Z]*[0-9]*[\\=*]*|[?*]\", string))\n",
    "    amino_acid = amino_acid.strip(\"[]''\")\n",
    "    amino_acid = amino_acid.replace(\"*\", \"Ter\")\n",
    "    amino_acid = amino_acid.replace(\"=\", amino_acid[:3])\n",
    "    return amino_acid\n",
    "\n",
    "# Accession number\n",
    "df['AccessionNo'] = df['cDNA nomenclature'].apply(lambda x: re.split(\"\\:\", str(x), 1)[0])\n",
    "# cDNA\n",
    "df['cDNANo'] = df['cDNA nomenclature'].apply(lambda x: str(re.findall(\"[^c\\.][0-9]+[+-_]*[0-9]+\", str(x))[-1:]).strip(\"[]''\"))\n",
    "# AminoAccid\n",
    "df['AminoNo'] = df['Protein nomenclature'].apply(lambda x: amino_acid(str(x)))\n",
    "# Genomic\n",
    "df['GenomicNo'] = df['Genomic nomenclature'].apply(lambda x: str(re.findall(\"[0-9]+[\\-]*[0-9]*\", str(x))[-1:]).strip(\"[]''\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Functions\n",
    "def gene_name(gene):\n",
    "    return re.split(\"\\_\", str(gene), 1)[0]\n",
    "\n",
    "def single_amino_code(aa3):\n",
    "    aa_dict = {'Ala': 'A', 'Arg': 'R', 'Asn': 'N', 'Asp': 'D', 'Cys':'C', 'Glu':'E','Gln':'Q', \\\n",
    "               'Gly':'G', 'His':'H', 'Ile':'I','Leu':'L','Lys':'K','Met':'M','Phe':'F','Pro':'P','Ser':'S', \\\n",
    "               'Thr':'T','Trp':'W','Tyr':'Y','Val':'V', 'Ter':'X','?':'?'}\n",
    "    return aa_dict[aa3]\n",
    "\n",
    "\n",
    "def zygosity(genotype): # need to do something with non-classic zygosity variants\n",
    "    try:\n",
    "        genotype = float(genotype)\n",
    "        if (genotype >= 0.4 and genotype <= 0.7) or (genotype >=1.4 and genotype <= 1.6):\n",
    "            return \"0/1\"\n",
    "        elif genotype <=0.1:\n",
    "            return \"1/1\"\n",
    "        else:\n",
    "            return genotype\n",
    "    except:\n",
    "        return genotype\n",
    "    \n",
    "    \n",
    "def get_exons(gene):\n",
    "    try:\n",
    "        gene = gene.split(\",\")\n",
    "        exons = [re.findall(\"_[0-9]+\", x)[0].replace(\"_\",\"\") for x in gene]\n",
    "    except:\n",
    "        return gene\n",
    "\n",
    "    if len(exons) == 1:\n",
    "        return int(exons[0])\n",
    "    else:\n",
    "        return exons[0] + \"-\" + exons[-1]\n",
    "    \n",
    "    \n",
    "def variant_type(varType, insBases):\n",
    "    if varType == \"duplication\":\n",
    "        return \"dup\"\n",
    "    elif varType == \"deletion\":\n",
    "        return \"del\"\n",
    "    elif varType == \"insertion\":\n",
    "        return \"ins\" + insBases\n",
    "    elif varType == \"delins\":\n",
    "        return \"delins\" + insBases\n",
    "    else:\n",
    "        return \"\"\n",
    "     \n",
    "\n",
    "def mutation_details(chrom, gene, exon, intron, amino, cdna, ref, alt, genomic, genotype, varType, insBases):\n",
    "    \n",
    "    if pd.isnull(intron):\n",
    "        coding = \"ex\"\n",
    "    else:\n",
    "        coding = \"int\"\n",
    "    \n",
    "    if pd.isnull(exon):\n",
    "        ex_int = get_exons(gene)\n",
    "    else:\n",
    "        try:\n",
    "            ex_int = int(exon)\n",
    "        except:\n",
    "            ex_int = np.nan\n",
    "    gene = gene_name(gene)\n",
    "    genotype = zygosity(genotype)\n",
    "    \n",
    "    varType = variant_type(varType, insBases)\n",
    "    \n",
    "    if pd.isnull(ref) and pd.isnull(exon):\n",
    "        nucleotide = f\"g.{genomic}{varType}\"\n",
    "    elif pd.isnull(ref):\n",
    "        nucleotide = f\"c.{cdna}{varType}\"\n",
    "    else:\n",
    "        nucleotide = f\"c.{cdna}{ref}>{alt}{varType}\"\n",
    "        \n",
    "    try:\n",
    "        amino = amino.replace(\"=\", amino[:3])\n",
    "        amino = re.sub(\"[a-zA-Z]{3}fs\\*\", \"fs\", amino)\n",
    "        amino = amino.replace(\"*\", \"Ter\")\n",
    "    except:\n",
    "        pass\n",
    "\n",
    "    if chrom == \"X\" or isinstance(genotype, float):\n",
    "        return f\"{gene} {coding}{ex_int} p.{amino} {nucleotide}\"\n",
    "    elif genotype == \"0/1\":\n",
    "        return f\"{gene} {coding}{ex_int} p.{amino}/N {nucleotide}/N\"\n",
    "    elif genotype == \"1/1\":\n",
    "        return f\"{gene} {coding}{ex_int} p.{amino}/{amino} {nucleotide}/{nucleotide[2:]}\"\n",
    "    \n",
    "    \n",
    "def mut_surveyor(reportVar, chrom, genotype, amino, nucleotide, genomic, ref, alt, insBases, varType):\n",
    "    if chrom == \"X\":\n",
    "        genotype = \"1/1\"\n",
    "    else:\n",
    "        genotype = zygosity(genotype)\n",
    "    \n",
    "    try:\n",
    "        if len(varType) != 6: varType = varType[:3] # need to check cases of delins\n",
    "    except:\n",
    "        pass\n",
    "    \n",
    "    try:\n",
    "        amino_sep = re.findall('[a-zA-Z]{3}|[\\d]+|[_=*]|fs*|fs|[?*]', amino)\n",
    "    except:\n",
    "        print(\"error:\", amino)\n",
    "        pass\n",
    "    \n",
    "    # determine if cDNA or genomic nomenclature should be used\n",
    "    if (nucleotide == \"\" and amino == \"\") and reportVar != \"No variant detected\":\n",
    "        genomic = genomic.split(\"-\")\n",
    "        nucleotide = genomic[0] + \"_\" + genomic[1]\n",
    "    \n",
    "    # final mutation surveyor output\n",
    "    if varType == \"sub\":\n",
    "        if genotype == \"0/1\":\n",
    "            try:\n",
    "                amino = single_amino_code(amino_sep[0])+amino_sep[1]+single_amino_code(amino_sep[0])+single_amino_code(amino_sep[2])\n",
    "                return f\"c.[{nucleotide}{ref}>{alt}]+[=],p.{amino}\"\n",
    "            except:\n",
    "                return f\"c.[{nucleotide}{ref}>{alt}]+[=],p.{amino}\"\n",
    "        elif genotype == \"1/1\":\n",
    "            try:\n",
    "                amino = single_amino_code(amino_sep[0])+amino_sep[1]+single_amino_code(amino_sep[2])\n",
    "                return f\"c.[{nucleotide}{ref}>{alt}]+[{nucleotide}{ref}>{alt}],p.{amino}\"\n",
    "            except:\n",
    "                return f\"c.[{nucleotide}{ref}>{alt}]+[{nucleotide}{ref}>{alt}],p.{amino}\"\n",
    "    # insertions\n",
    "    elif varType == \"ins\":\n",
    "        if genotype == \"0/1\":\n",
    "            return f\"c.[{nucleotide}_hetins{insBases}\"\n",
    "        elif genotype == \"1/1\":\n",
    "            return f\"c.[{nucleotide}_ins{insBases}\"\n",
    "    # delins\n",
    "    elif varType == \"delins\":\n",
    "        if genotype == \"0/1\":\n",
    "            return f\"c.[{nucleotide}_hetdelins{insBases}\"\n",
    "        elif genotype == \"1/1\":\n",
    "            return f\"c.[{nucleotide}_delins{insBases}\"\n",
    "    #structural variants\n",
    "    elif varType == \"del\" or varType == \"dup\":\n",
    "        if zygosity(genotype) == \"0/1\":\n",
    "            return f\"c.{nucleotide}_het{varType}\"\n",
    "        elif zygosity(genotype) == \"1/1\":\n",
    "            return f\"c.{nucleotide}_{varType}\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "df['MutDetails'] = df.apply(lambda x: mutation_details(x.Chromosome, x.Gene, x.Exon, x.Intron, x.AminoNo, x.cDNANo, x.Ref, x.Alt, x.GenomicNo, x.Genotype, x.varType, x.insBases), axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "df['MutSurveyor'] = df.apply(lambda x: mut_surveyor(x.reportVar, x.Chromosome, x.Genotype, x.AminoNo, x.cDNANo, x.GenomicNo, x.Ref, x.Alt, x.insBases, x.varType), axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = df[df['Folder number'].str.startswith(\"EX\")]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "export_csv('../output/', 'tNGS_export.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
